File: src/sor_seq.rs
use std::{env, time};
use std::f64::consts::PI;
use std::time::Instant;

fn stencil(matrix: &Vec<Vec<f64>>, row: usize, col: usize) -> f64 {
    (matrix[row - 1][col] + matrix[row + 1][col] + matrix[row][col - 1] + matrix[row][col + 1]) / 4.0
}

fn even_1_odd_0(num: usize) -> usize {
    match num % 2 {
        0 => 1,
        _ => 0
    }
}
pub fn sor(problem_size: usize, iteration_limit: usize) {
    // Problem size, default 1K (the matrix size)
    let mut N: usize  = problem_size;
    if N == 0 {
        N = 1000;
    }

    println!("Running SOR with {} rows", N);

    // Adding two borderlines
    N += 2;

    // Set up matrix info
    let n_col = N;
    let n_row = N;

    // Initialize parameters
    let r = 0.5 * ((PI / n_col as f64).cos() + (PI / n_row as f64).cos());
    let mut  omega =  2.0 / (1.0 + (1.0 - r * r).sqrt());
    let stop_diff = 0.0002 / (2.0 - omega);
    let mut max_diff;
    let mut diff;
    omega *= 0.8;

    // Initialize the matrix
    let mut mat = vec![vec![0.0; n_col]; n_row];
    let mut m = [[0.0; 1000]; 1000];
    let mut matrix: Vec<Vec<f64>> = Vec::with_capacity(n_row);
    for _ in 0..n_row {
        matrix.push(vec![0.0; n_col]);
    }
    for i in 0..n_row {
        for j in 0..n_col {
            if i == 0 {
                matrix[i][j] = 4.56;
            } else if i == n_row - 1 {
                matrix[i][j] = 9.85;
            } else if j == 0 {
                matrix[i][j] = 7.32;
            } else if j == n_col - 1 {
                matrix[i][j] = 6.88;
            } else {
                matrix[i][j] = 0.0;
            }
        }
    }

    // Now do the real computation
    let start = Instant::now();
    let mut iteration: i32 = 0;
    loop {
        max_diff = 0.0;
        for phase in 0..2 {
            for i in 1..(N-1) {
                let start_col = 1 + (even_1_odd_0(i) ^ phase);
                for j in (start_col..(N-1)).step_by(2) {
                    let stencil_val = stencil(&matrix, i, j);
                    diff = (stencil_val - matrix[i][j]).abs();
                    if diff > max_diff {
                        max_diff = diff;
                    }
                    matrix[i][j] = matrix[i][j] + omega * (stencil_val - matrix[i][j]);
                }
            }
        }
        iteration += 1;
        if max_diff <= stop_diff {
            break;
        }
    }
    let elapsed = start.elapsed();

    // Print results
    println!("SOR size: {} x {}, time: {} ms", n_row-2, n_col-2, elapsed.as_millis());
    println!("Using {} iterations, diff is {} (allowed diff {})", iteration, max_diff, stop_diff);
}

File: sor-seq.c
/*
 * Successive over relaxation
 * (red-black SOR)
 *
 */

#include "stdio.h"
#include "stdlib.h"
#include "math.h"
#include <sys/time.h>

int even (int i){
  return ( ( ( i / 2 ) * 2 ) == i ) ? 1 : 0;
}

double abs_d (double d){
  return (d < 0.0) ? -d : d;
}

double stencil (double**G, int row, int col){
  return ( G[row-1][col] + G[row+1][col] + G[row][col-1] + G[row][col+1] )
         / 4.0;
}


int main (int argc, char *argv[]){

  int N;                            /* problem size */
  int ncol,nrow;                    /* number of rows and columns */
  double TOLERANCE = 0.0002;        /* termination criterion */
  double Gnew,r,omega,              /* some float values */
         stopdiff,maxdiff,diff;     /* differences btw grid points in iters */
  double **G;                       /* the grid */
  int i,j,phase,iteration;          /* counters */

  /* set up problem size */
  N = 0;
  if ( argc > 1 )
    N = atoi(argv[1]);
  if (N == 0)
    N = 1000;


  printf("Running SOR with %d rows\n", N);

  N += 2; /* add the two border lines */
          /* finally N*N (from argv) array points will be computed */

  /* set up a quadratic grid */
  ncol = nrow = N;
  r        = 0.5 * ( cos( M_PI / ncol ) + cos( M_PI / nrow ) );
  omega    = 2.0 / ( 1 + sqrt( 1 - r * r ) );
  stopdiff = TOLERANCE / ( 2.0 - omega );
  omega   *= 0.8;                   /* magic factor */

  G = (double**)malloc(N*sizeof(double*));
  if ( G == 0 ){ printf("malloc failed\n"); exit(42); }
  for (i = 0; i<N; i++){ /* malloc the own range plus one more line */
                         /* of overlap on each side */
    G[i] = (double*)malloc(N*sizeof(double));
    if ( G[i] == 0 ){ printf("malloc failed\n"); exit(42); }
  }

  /* initialize the grid */
  for (i = 0; i < N; i++){
    for (j = 0; j < N; j++){
      if (i == 0) G[i][j] = 4.56;
      else if (i == N-1) G[i][j] = 9.85;
      else if (j == 0) G[i][j] = 7.32;
      else if (j == N-1) G[i][j] = 6.88;
      else G[i][j] = 0.0;
    }
  }

  struct timeval start, end;
  long mtime, seconds, useconds;
  /* now do the "real" computation */
  iteration = 0;
  gettimeofday(&start, NULL);
  do {
    maxdiff = 0.0;
    // [0, 1]
    for ( phase = 0; phase < 2 ; phase++){
      // [1, second last element]
      for ( i = 1 ; i < N-1 ; i++ ){
        // 1 + (even(1) ^ 1) => 1 + 1 = 2; j => 4, 6, 
        // 1 + (even(1) ^ 0) => 1 + 0 = 1; j => 3
        // Phase 0: j is odd, Phase 1: j is even
        for ( j = 1 + (even(i) ^ phase); j < N-1 ; j += 2 ){
          Gnew = stencil(G,i,j); // average of four direction
          diff = abs_d(Gnew - G[i][j]);
          if ( diff > maxdiff )
            maxdiff = diff; // enlarge maxDiff
          G[i][j] = G[i][j] + omega * (Gnew-G[i][j]); // update current cell value
        }
      }
    }
    iteration++;
  } while (maxdiff > stopdiff);

    gettimeofday(&end, NULL);
    seconds  = end.tv_sec  - start.tv_sec;
    useconds = end.tv_usec - start.tv_usec;
    mtime = ((seconds) * 1000 + useconds/1000.0) + 0.5;

    printf("SOR %d x %d complete, took %ld ms to complete\n",N-2,N-2, mtime);
    printf("using %d iterations, diff is %f (allowed diff %f)\n",
           iteration,maxdiff,stopdiff);

  return 0;
}
